### PR TITLE
Removing usage of CancellationToken.Register

### PR DESCRIPTION
**Pull request by sburmanoctopus**
[sc-58580]

# Background

During Halibut testing, [it was found](https://github.com/OctopusDeploy/Halibut/pull/481) that using `CancellationToken.Register` could cause unwanted memory leaks, and odd patterns when analysing memory.

Therefore, we decided to remove the usage of `CancellationToken.Register`.

# Results
Related to OctopusDeploy/Issues#8229

## Before

We would use `CancellationToken.Register` to see if a task was abandoned.

## After

We now achieve the same outcome using `Task.Delay` and `CancellationTokenSource`.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs [hunk 1]
```diff
@@ -23,5 +25,43 @@ public static void IgnoreUnobservedExceptions(this Task task)
...
                 TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously);
         }
+
+        public static async Task<TaskCompletionResult> WaitTillCompletion(this Task actionTask, TimeSpan abandonAfter, CancellationToken cancellationToken)
+        {
+            var actionTaskCompleted = await actionTask.WaitTillCompletedOrCancelled(cancellationToken);
+
+            if (actionTaskCompleted) return TaskCompletionResult.Completed;
+
+            actionTaskCompleted = await actionTask.WaitTillCompletedOrTimeout(abandonAfter).ConfigureAwait(false);
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs:35]: nit ConfigureAwait is sometimes used. Should we go consistent and use it or don't

File: source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs [hunk 2]
```diff
@@ -23,5 +25,43 @@ public static void IgnoreUnobservedExceptions(this Task task)
...
+
+            if (actionTaskCompleted) return TaskCompletionResult.Completed;
+
+            actionTask.IgnoreUnobservedExceptions();
+            return TaskCompletionResult.Abandoned;
+
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs:41]: nit: space

**Review by nathanwoctopusdeploy — APPROVED** [review]
LGTM

File: source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs (hunk 2 above)
Comment (sburmanoctopus) [reply source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs:41]: Removed

File: source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs (hunk 1 above)
Comment (sburmanoctopus) [reply source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs:35]: Yeah, I'm still not 100% sure what our policy on this is.

I think that we do not use it, right?

Either way, I've removed `ConfigureAwait` from this file only.

Comment (nathanwoctopusdeploy) [reply source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs:35]: Yeh I think we went with we don't need it

**Pull request merged**